### PR TITLE
storage: remove preemptive snapshots

### PR DESCRIPTION
**Pull request by ajwerner**
This commit adds a migration to remove any remaining on-disk pre-emptive
snapshots and then excises the concept from the codebase.

There's some cruft in replica_init that I'd love to clean up but don't
have a clear vision of how.

Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
@tbg here's a prototype of removing preemptive snapshots. It still needs testing.

One thing I tried to disambiguate was the word "initialized" as it applies to the `Replica`. Confusingly, calls to `init` did not previously always result in an "initialized" `Replica`. I carved out instead a notion of "loaded" as distinct from initialized. I sort of hate that that concept needs to exist at all tbh. 

I'd love to hear your thoughts on further improvements. Thanks!

File: pkg/storage/replica_command.go [hunk 1]
```diff
@@ -1971,8 +1871,12 @@ func (r *Replica) sendSnapshot(
 			},
 		},
 		RangeSize: r.GetMVCCStats().Total(),
-		// Recipients can choose to decline preemptive snapshots.
-		CanDecline: snapType == SnapshotRequest_PREEMPTIVE,
+		// Recipients currently cannot choose to decline any snapshots.
+		// TODO(ajwerner): Should a recipient be able to decline a LEARNER snapshot?
+		// It seems like yes. The recipient may know that it does not want to get
+		// added to a range. Confer with danhhz/tbg/darin about the implications of
+		// making this true for SnapshotRequest_LEARNER.
```
Comment (tbg) [inline pkg/storage/replica_command.go]: This was just discussed on https://github.com/cockroachdb/cockroach/pull/44247#issuecomment-577668089

TLDR: when you refuse the learner, you've already added it to the desc, so we need to be careful. We don't want to roll back replication changes en masse which is what would happen here.

File: pkg/storage/store.go [hunk 2]
```diff
@@ -1332,8 +1334,16 @@ func (s *Store) Start(ctx context.Context, stopper *stop.Stopper) error {
 			if !desc.IsInitialized() {
 				return false, errors.Errorf("found uninitialized RangeDescriptor: %+v", desc)
 			}
+			replicaDesc, found := desc.GetReplicaDescriptor(s.StoreID())
+			if !found {
```
Comment (tbg) [inline pkg/storage/store.go]: We also get replicas like that when a replica catches up across a change replicas trigger that removes it. When you made these replicas GC themselves proactively when they hit the trigger, you did so atomically with the application of the CRT raft command, right? In other words, in this branch, are we really 100% looking at a preemptive snap? I suppose at the very least we might be looking at a replica that did catch up across the trigger before we had code that insta-gc'ed, but it seems fine to treat that as a preemptive snapshot and nuke it (which would happen to it eventually anyway, since it's not in the desc any more).

File: pkg/storage/store_create_replica.go [hunk 3]
```diff
@@ -132,27 +121,21 @@ func (s *Store) tryGetOrCreateReplica(
 		// If this is intended for replicaID 0 then it's either a preemptive
 		// snapshot or a split/merge lock in which case we'll let it go through.
 		if replicaID == 0 {
+			// TODO(ajwerner): fix this case.
```
Comment (tbg) [inline pkg/storage/store_create_replica.go]: I have this all paged out again, but is the split/merge lock the last remaining use of replicaID zero? I'd be so happy to not have that be a thing. (not in this PR)

**Comment by tbg** [comment]
Only a very partial review, I will have to give this a look with fresh eyes next week. Excited that this is happening!

File: pkg/storage/store_create_replica.go (hunk 3 above)
Comment (ajwerner) [reply pkg/storage/store_create_replica.go]: Even better! I removed all uses of ReplicaID 0 in this PR already, just forgot to come back to this conditional after I did that. This can be a fatal now.

Comment (tbg) [reply pkg/storage/store_create_replica.go]: ❤️

**Comment by ajwerner** [comment]
This is RFAL. I'll put up a second PR to update the grafana dashboards.

**Review by tbg — APPROVED** [review]
:lgtm_strong: : very very nice. I'm so happy you chewed through and removed all of that detritus. I feel a lot better about all of the code you touched.

> Reviewed 14 of 25 files at r1, 14 of 14 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-:-M-EW3Vf7D9xOi5hrIIC:b-4o0biw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/storage/replica_init.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EPx9F1eUYtZfEpX4C:-M-EPx9F1eUYtZfEpX4D:b8cpoob) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/replica_init.go#L278)):*
> ```Go
> 	//      batch. We could avoid setting the local descriptor in this case.
> 	//   2) When the DisableEagerReplicaRemoval testing knob is enabled. We
> 	//      could
> ```

missing some words

I agree that it would be nice to be more coherent here. Not necessarily for this PR, looks like a decent amount of individual cleanup is needed.
___
*[pkg/storage/replica_raft.go, line 1538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EQfaM0tSz7ZYYQWD8:-M-EQfaM0tSz7ZYYQWD9:b-y5xfr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/replica_raft.go#L1538)):*
> ```Go
> 	ctx context.Context, split *roachpb.SplitTrigger,
> ) (func(), error) {
> 	// We pass a 0 replicaID because we want to lock the RHS even if we know
> ```

Comment stale.
___
*[pkg/storage/split_queue_test.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EQzK_1QQRTuNIHN1x:-M-EQzK_1QQRTuNIHN1y:b-7qgczx) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/split_queue_test.go#L77)):*
> ```Go
> 		// ensures that the store won't be mucking with our replica concurrently
> 		// during testing (e.g. via the system config gossip update).
> 		copy := *tc.repl.Desc()
> ```

Shadowing a built-in? I usually go with `cpy` in that case.
___
*[pkg/storage/store.go, line 1342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ER8OU90LAzUfEkIb3:-M-ER8OU90LAzUfEkIb4:b-51mmqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store.go#L1342)):*
> ```Go
> 				// range which has processed a raft command to remove itself (which is
> 				// possible prior to 19.2 or if the DisableEagerReplicaRemoval is
> 				// enabled) and has not yet been removed by the replica	 gc queue.
> ```

weird spaces
___
*[pkg/storage/store_split.go, line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ESiAd1A37Xxsq_UQm:-M-ESiAd1A37Xxsq_UQn:b-sb7u1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L51)):*
> ... if it hasn't been removed in the meantime (handled below).
___
*[pkg/storage/store_split.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ESJFV2ORp7Y0LTvkN:-M-ESJFV2ORp7Y0LTvkO:bi116rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L53)):*
> ```Go
> 	// acquired) in Replica.acquireSplitLock. It must be present here.
> 	rightRepl, err := r.store.GetReplica(split.RightDesc.RangeID)
> 	if err != nil && !roachpb.IsRangeNotFoundError(err) {
> ```

I'd rewrite this for clarity as

```go
rightRepl, err := r.store.GetReplica(split.RightDesc.RangeID)
if roachpb.IsRangeNotFoundError(err) {
  // The right hand side we were planning to populate has already been removed.
  // We handle this below.
  rightRepl = nil
}
if err != nil {
  log.Fatalf(ctx, errors.Wrap(err, "failed to get RHS replica"))
}
```
___
*[pkg/storage/store_split.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ET7ItAieQUgVZ9RIU:-M-ET7ItAieQUgVZ9RIV:buyz9bf) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L62)):*
> ```Go
> 		// We're in the rare case where we know that the RHS has been removed
> 		// and re-added with a higher replica ID (and then maybe removed again).
> 		// We know we've never processed a snapshot for the right range because up
> ```

Worth an assertion (if rightRng isn't nil).
___
*[pkg/storage/store_split.go, line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ESxu3C2jQd9kWcHH2:-M-ESxu41jRsAj8L8rhr:b-kjrh5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L63)):*
> ```Go
> 		// and re-added with a higher replica ID (and then maybe removed again).
> 		// We know we've never processed a snapshot for the right range because up
> 		// to this point it would overlap	with the left and ranges cannot move
> ```

spaces
___
*[pkg/storage/store_split.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ETbpQ40I4uoSH2grL:-M-ETbpQ40I4uoSH2grM:b-u8mqoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L66)):*
While you're here, rework this comment - it's unclear what it's about until one hits this paragraph. Something like

> // We're in the rare case where we know that the RHS has been removed
// and re-added with a higher replica ID (and then maybe removed again).
//
// To apply the split, we need to "throw away" the data that would belong to
// the RHS, i.e. we clear the user data the RHS would have inherited from the
// LHS due to the split and additionally clear all of the range ID local state
// that the split trigger writes into the RHS.
//
// We know we've never processed a snapshot for the right range because the
// LHS prevents any incoming snapshots until the split has executed (i.e. now).
// It is important to preserve the HardState because we might however have
// already voted at a higher term. In general this shouldn't happen because
// we add learners and then promote them only after we snapshot but we're
// going to be extra careful in case future versions of cockroach somehow
// promote replicas without ensuring that a snapshot has been received.
//
// Rather than specifically deleting around the data we want to preserve
// we read the HardState to preserve it, clear everything and write back
// the HardState and tombstone. Note that we only do this if rightRepl
// exists; if it doesn't, there's no Raft state to massage (when rightRepl
// was removed, a tombstone was written instead).
___
*[pkg/storage/store_split.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EV4Gb1t_qj4tPieXB:-M-EV4Gb1t_qj4tPieXC:b-3tcgqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L111)):*
> ```Go
> // split with the Store. Requires that Replica.raftMu is held.
> //
> // TODO(tschottdorf): Want to merge this with SplitRange, but some legacy
> ```

Remove this TODO.
___
*[pkg/storage/store_split.go, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EV8XxEBuz7ArA5sWD:-M-EV8XxEBuz7ArA5sWE:b-trxfxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L116)):*
> ```Go
> 	ctx context.Context, deltaMS enginepb.MVCCStats, split *roachpb.SplitTrigger, r *Replica,
> ) {
> 	rightRngOrNil := prepareRightReplicaForSplit(ctx, split, r)
> ```

Mention when it's nil.
___
*[pkg/storage/store_split.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EVH9g8GDeCoaHkuhr:-M-EVH9g8GDeCoaHkuhs:b-fzj89u) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L163)):*
> ```Go
> 	// If the RHS replica at the point of the split was known to be removed when
> 	// during the application of the split then we may not find it here. That's
> 	// fine, carry on.
> ```

```go
// See also:
_, _= r.acquireSplitLock, r.splitPreApply
```

**Review by ajwerner — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-:-M-FtjAw2JoTE1Ort0_z:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/replica_command.go, line 1878 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-LzxrMzOt8-mA1i9VxNA-r1-1878:-M-FnGbU9GILqXQjwB2C:bb6sh2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef1de4ec18ceffac257aca30e7469a5756848b1d/pkg/storage/replica_command.go#L1878)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This was just discussed on https://github.com/cockroachdb/cockroach/pull/44247#issuecomment-577668089

TLDR: when you refuse the learner, you've already added it to the desc, so we need to be careful. We don't want to roll back replication changes en masse which is what would happen here.
</blockquote></details>

Ack.
___
*[pkg/storage/replica_init.go, line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EPx9F1eUYtZfEpX4C:-M-FswvbEiBaDbvqJ2EG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/replica_init.go#L278)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

missing some words

I agree that it would be nice to be more coherent here. Not necessarily for this PR, looks like a decent amount of individual cleanup is needed.
</blockquote></details>

Done.
___
*[pkg/storage/replica_raft.go, line 1538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EQfaM0tSz7ZYYQWD8:-M-FtdeL2NwmJnIbiTIR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/replica_raft.go#L1538)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Comment stale.
</blockquote></details>

Done.
___
*[pkg/storage/split_queue_test.go, line 77 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EQzK_1QQRTuNIHN1x:-M-FoJy0EiyPckRRq8CG:b2asrjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/split_queue_test.go#L77)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Shadowing a built-in? I usually go with `cpy` in that case.
</blockquote></details>

Done.  This was ancient code. 
___
*[pkg/storage/store.go, line 1342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ER8OU90LAzUfEkIb3:-M-FoZ_yDsYECFy7g_C5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store.go#L1342)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

weird spaces
</blockquote></details>

Done.
___
*[pkg/storage/store_split.go, line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ESiAd1A37Xxsq_UQm:-M-FojOx6BmUeRVI5cW3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L51)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

> ... if it hasn't been removed in the meantime (handled below).
</blockquote></details>

Done.
___
*[pkg/storage/store_split.go, line 53 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ESJFV2ORp7Y0LTvkN:-M-Fp3V4-B3VlaQE-xX1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L53)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I'd rewrite this for clarity as

```go
rightRepl, err := r.store.GetReplica(split.RightDesc.RangeID)
if roachpb.IsRangeNotFoundError(err) {
  // The right hand side we were planning to populate has already been removed.
  // We handle this below.
  rightRepl = nil
}
if err != nil {
  log.Fatalf(ctx, errors.Wrap(err, "failed to get RHS replica"))
}
```
</blockquote></details>

Done.
___
*[pkg/storage/store_split.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ET7ItAieQUgVZ9RIU:-M-FqpGMDAOMhixlILCr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L62)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Worth an assertion (if rightRng isn't nil).
</blockquote></details>

Done.
___
*[pkg/storage/store_split.go, line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ESxu3C2jQd9kWcHH2:-M-Fqp26Dbbd-tjzwi4l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L63)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

spaces
</blockquote></details>

Done.
___
*[pkg/storage/store_split.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-ETbpQ40I4uoSH2grL:-M-FqqDRBt0dzrq7oxFa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L66)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

While you're here, rework this comment - it's unclear what it's about until one hits this paragraph. Something like

> // We're in the rare case where we know that the RHS has been removed
// and re-added with a higher replica ID (and then maybe removed again).
//
// To apply the split, we need to "throw away" the data that would belong to
// the RHS, i.e. we clear the user data the RHS would have inherited from the
// LHS due to the split and additionally clear all of the range ID local state
// that the split trigger writes into the RHS.
//
// We know we've never processed a snapshot for the right range because the
// LHS prevents any incoming snapshots until the split has executed (i.e. now).
// It is important to preserve the HardState because we might however have
// already voted at a higher term. In general this shouldn't happen because
// we add learners and then promote them only after we snapshot but we're
// going to be extra careful in case future versions of cockroach somehow
// promote replicas without ensuring that a snapshot has been received.
//
// Rather than specifically deleting around the data we want to preserve
// we read the HardState to preserve it, clear everything and write back
// the HardState and tombstone. Note that we only do this if rightRepl
// exists; if it doesn't, there's no Raft state to massage (when rightRepl
// was removed, a tombstone was written instead).


</blockquote></details>

Done.
___
*[pkg/storage/store_split.go, line 111 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EV4Gb1t_qj4tPieXB:-M-FqyMUC9_ouTXQR_Ht:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L111)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Remove this TODO.
</blockquote></details>

Done.
___
*[pkg/storage/store_split.go, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EV8XxEBuz7ArA5sWD:-M-Fr_yc9ED_-yIorkQF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L116)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Mention when it's nil.
</blockquote></details>

Done.
___
*[pkg/storage/store_split.go, line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-EVH9g8GDeCoaHkuhr:-M-FrsX53idVTVCM8sdF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df229154e35bae82350513607503cea127a44a76/pkg/storage/store_split.go#L163)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

```go
// See also:
_, _= r.acquireSplitLock, r.splitPreApply
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

**Review by tbg — APPROVED** [review]
:lgtm:

> Reviewed 11 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-:-M-G7CvD5crv1zwLWCkZ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/storage/store_split.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-G71sZB8BZHUeegMog:-M-G71sZB8BZHUeegMoh:b-hb6sk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/454296b9e97960c299f116bd666b10b40f1c13ba/pkg/storage/store_split.go#L91)):*
> ```Go
> 			// snapshot to initialize this newer RHS.
> 			if rightRepl.IsInitialized() {
> 				log.Fatalf(ctx, "cannot delete ")
> ```

not that we're going to ever see this fire :crossed_fingers: but something's missing here

**Review by ajwerner — COMMENTED** [review]
TFTR!

bors r=tbg

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-:-M-GJdgJExj1gQwpdoWJ:b-kg5fnc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/store_split.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44615#-M-G71sZB8BZHUeegMog:-M-GJX2X5rL4XrcdYl6S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/454296b9e97960c299f116bd666b10b40f1c13ba/pkg/storage/store_split.go#L91)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

not that we're going to ever see this fire :crossed_fingers: but something's missing here
</blockquote></details>

Done.
